File: 10 Days of Statistics/Day 7: Pearson Correlation Coefficient I.py
# https://www.hackerrank.com/challenges/s10-pearson-correlation-coefficient/problem

# use of scipy/numpy not llowed

"""
Objective
In this challenge, we practice calculating the Pearson correlation coefficient.

Task
Given two n-element data sets, X and Y, calculate the value of the Pearson correlation coefficient.
"""

from math import sqrt

# function to calculate the Covariance, both lists should be of same length
# Σ( (xi-µx)*(yi-µy) ) / n
def covariance(X,Y):
    n = len(X)
    X_mean = sum(X)/n
    Y_mean = sum(Y)/n
    return ( sum( [ ( (x-X_mean) * (y-Y_mean) ) for x, y in zip(X, Y) ] ) / n )

# function to calculate the average squared distance from the mean, the Variance
def variance(ls_val):
    n = len(ls_val)
    ls_val_avg = sum(ls_val)/n
    sum_sq_dist = 0
    for i in ls_val:
        # substract the avg from the i value and sum to the total squared distance
        sum_sq_dist +=(i-ls_val_avg)**2
    #devide the sum of the squared distances by the count of elements in the list
    return  sum_sq_dist/len(ls_val)

# function to calculate the standard deviation using the variance
def sd(ls_val):
    return sqrt(variance(ls_val))

# function to calculate the pearson correlation coefficient
# covariance / (σx*σy)
def pcc(X,Y):
    return covariance(X,Y) / ( sd(X) * sd(Y) )

# size of data sets X and Y
# number of elements computed using len()
# following variable only needed for the challenge to work but not for the script
n = float(input().strip())

# data set X
X = list(map(float, input().strip().split()))

# data set Y
Y = list(map(float, input().strip().split()))

print(round(pcc(X,Y),3))

File: 10 Days of Statistics/Day 7: Pearson Correlation Coefficient I.R
# https://www.hackerrank.com/challenges/s10-pearson-correlation-coefficient/problem

# built-in function cov() is for sample not population
# the build in sd() function is for sample not population
# an option -> sqrt((n-1)/n) * sd(x)

# function to calculate the Covariance, both lists should be of same length
# Σ( (xi-µx)*(yi-µy) ) / n
covariance <- function(X,Y){
    n <- length(X)
    X_min_mean <- X - mean(X)
    Y_min_mean <- Y - mean(Y)
    sum( (X_min_mean) * (Y_min_mean) ) / n 
}

# function to calculate the population sd
popsd <- function(v_val){
    # average the elements in the vector
    v_val_avg <- mean(v_val)
    # squared distance from the mean
    v_val_dist <- (v_val-v_val_avg)^2
    # average squared distance from the mean
    v_val_dist_avg <- sum(v_val_dist) / length(v_val)
    # SD by square root of average squared distance from the mean (variance)
    sqrt(v_val_dist_avg)
}   

# function to calculate the pearson correlation coefficient
# covariance / (σx*σy)
pcc <- function(X,Y){
    covariance(X,Y) / ( popsd(X) * popsd(Y) )
}

con <- file("stdin", open = "r")

dataset <- readLines(con)

# size of data sets X and Y
# following variable not used, number of elements computed using length 
n <- as.numeric(dataset[1])

# data set X
X <- unlist(strsplit(dataset[2], split=" "))
X <- as.numeric(X)

# data set Y
Y <- unlist(strsplit(dataset[3], split=" "))
Y <- as.numeric(Y)

cat(round(pcc(X,Y),digits=3))
